test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,443 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:generate protoc --go_out=. trf.proto
+package bpftests
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/perf"
+	"github.com/golang/protobuf/proto"
+	"github.com/vishvananda/netlink/nl"
+	"google.golang.org/protobuf/encoding/protowire"
+
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/datapath/link"
+	"github.com/cilium/cilium/pkg/monitor"
+)
+
+var testPath = flag.String("bpf-test-path", "", "Path to the eBPF tests")
+
+func TestBPF(t *testing.T) {
+	if testPath == nil || *testPath == "" {
+		t.Fatal("-bpf-test-path is a required flag")
+	}
+
+	entries, err := os.ReadDir(*testPath)
+	if err != nil {
+		t.Fatal("os readdir: ", err)
+	}
+
+	pinPath := "/sys/fs/bpf/cilium-test"
+	defer func() {
+		os.RemoveAll(pinPath)
+	}()
+	if err = os.Mkdir(pinPath, 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+
+		if !strings.HasSuffix(entry.Name(), ".o") {
+			continue
+		}
+
+		func() {
+			elfPath := path.Join(*testPath, entry.Name())
+			spec, err := ebpf.LoadCollectionSpec(elfPath)
+			if err != nil {
+				t.Fatal("load spec: ", elfPath, err)
+			}
+
+			for _, mspec := range spec.Maps {
+				// Drain extra info from map specs
+				if mspec.Extra != nil {
+					io.ReadAll(mspec.Extra)
+				}
+
+				// Remove BTF from maps as a workaround since some maps, like 'cilium_xdp_scratch' will return a
+				// "map create: invalid argument" error if we don't
+				mspec.BTF = nil
+			}
+
+			// Detect program type mismatches
+			var progTestType ebpf.ProgramType
+			for _, spec := range spec.Programs {
+				if progTestType != spec.Type {
+					if progTestType == ebpf.UnspecifiedProgram {
+						progTestType = spec.Type
+						continue
+					}
+					if spec.Type == ebpf.UnspecifiedProgram {
+						continue
+					}
+
+					t.Fatalf(
+						"File '%s' contains both '%s' and '%s' program types, "+
+							"only one program type per ELF file allowed:",
+						elfPath,
+						progTestType,
+						spec.Type,
+					)
+				}
+			}
+
+			if progTestType == ebpf.UnspecifiedProgram {
+				t.Fatalf("File '%s' only contains unspecified program types", elfPath)
+			}
+
+			// Give all tail call programs the same program type as the test programs
+			for _, spec := range spec.Programs {
+				spec.Type = progTestType
+			}
+
+			coll, err := ebpf.NewCollectionWithOptions(spec, ebpf.CollectionOptions{
+				Programs: ebpf.ProgramOptions{
+					LogSize: 1 << 20,
```

Comment:
Agreed on both counts, creating a loop to grow the log size if ENOSPC is encountered is trivial, have done this before. Then again, do we want more than 1MB of verifier log in the CI output? Perhaps we can do something here with flags for size and verbosity so we don't get that much output in the CI, but we can get detailed info when running locally to fix a detected issue.

As for how to handle this in general... It depends. It would be nice if we could incorporate this behavior in cilium/ebpf by specifying size -1 or something like it. But it is not something you want all the time, because of memory consumption and such.